### PR TITLE
DRAFT - Ruby work to role individual components

### PR DESCRIPTION
**Pull request by j-maynard**
This is a bit of a big change.  I have removed the following scripts as they don't seem to working any more:
* `restart.rb`
* `shutdown-jars.sh`

Updated the shutdown.sh script to actually shutdown dozzle if its running and added an option to override this behaviour.

Created a new shell script for component management to wrap originally around `build-local.rb` but now wraps the new `component.rb` script (see below).

Broke apart the `build-local.rb` script to be easier to maintain.  First I moved the class files in to their own files.  Next I separated out the new code for managing components in to a `components.rb` script.  Finally I did a bit of a clean up and moved all the ruby files in to the `lib` directory.  You shouldn't need to access these directly as we have wrapper shell scripts for this.

### CODE REVIEW

File: build-local.rb [hunk 1]
```diff
@@ -237,8 +245,75 @@ def generate_env(images)
...
+  output = `docker-compose --env-file .env up -d #{component_name} 2>&1`
+  if $?.success?
+    spinner.success(" - Service #{component_name} has been started.")
+  else
+    spinner.error(" - Could not start service #{component_name}")
+    puts "Debug information:\n\tComponent_name = #{component_name}\n\tDocker Output = #{output}"
+  end
+end
+
+def role_component(repos, component_name, skip_build = false)
```
Comment (JohnWattsGDS) [inline build-local.rb:305]:
```suggestion
def roll_component(repos, component_name, skip_build = false)
```

File: build-local.rb [hunk 2]
```diff
@@ -237,8 +245,75 @@ def generate_env(images)
   print "     Done\n"
 end
 
+def docker_ps
+  output = `docker ps --format '"{{$v1 := split .Image ":"}}{{$shortName := index $v1 0}}{{$shortName}}": { "name": "{{.Names}}", "image": "{{.Image}}", "container": "{{.ID}}" },'`
+  JSON.parse("{ #{output[0..-3]} }")
+end
+
+def list_compoents(repos)
```
Comment (JohnWattsGDS) [inline build-local.rb:253]: There's a few `compoents` that have slipped in so it might be worth running a find and replace and `s/compoents/components/g`

Comment (j-maynard) [reply build-local.rb:253]: I was trying to decide if that was the right name or if I should go for service?

Comment (JohnWattsGDS) [reply build-local.rb:253]: Either works for me. I think we have a kill-all-hub-services.sh in ida-hub-acceptance-tests if you wanted to be consistent with that.

File: build-local.rb [hunk 3]
```diff
@@ -4,6 +4,8 @@
 require 'os'
 require 'tty-spinner'
 require 'yaml'
+require 'json'
+require 'pry'
```
Comment (JohnWattsGDS) [inline build-local.rb:8]: Are we able to drop `pry` now?

Comment (j-maynard) [reply build-local.rb:8]: I'll drop it when the PR moves out of draft :-D 

Having said that pry is mightly useful especially when trying new things or developing and I keep adding it in for Dev and removing it when the PR goes live, part of me would like to stop 🛑 having to do that cycle ;-)

File: README.md [hunk 4]
```diff
@@ -64,6 +65,10 @@ If everything works as expected you should see the following message:
 
 It can take up to 5 minutes for the 13 individual microservices to get going and for you to make a successful verify journey depending on the age and power of your computer.  If you encounter any issues please [raise an issue](#Support-and-raising-issues).
 
+## Managing Components
+
+
```
Comment (instantiator) [inline README.md:70]: Feel free to update this in a second commit / additional PR.

File: lib/build-local.rb [hunk 5]
```diff
@@ -4,6 +4,12 @@
 require 'os'
 require 'tty-spinner'
 require 'yaml'
+require 'json'
+require 'colorize'
+require 'pry'
```
Comment (j-maynard) [inline lib/build-local.rb:9]: Remove this!

File: lib/build-local.rb [hunk 6]
```diff
@@ -20,127 +26,29 @@
 USAGE = <<ENDUSAGE
 Usage:
     build-local -y yaml-file [-t count] [-r retries] [-h]
```
Comment (j-maynard) [inline lib/build-local.rb:28]: Update -r to -R here!

File: lib/components.rb [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+    if $?.success?
+      spinner.success
+    else
+      spinner.error("Could not stop service #{component_name.bold}")
+      puts "Debug information:\n\tComponent_name = #{component_name}\n\tComponent = #{component}"
+    end
+  end
+end
+
+def rebild_component(repos, component_name)
```
Comment (j-maynard) [inline lib/components.rb:92]: should be rebuild

File: lib/components.rb [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+  output = `docker-compose --env-file .env up -d #{component_name} 2>&1`
+  if $?.success?
+    spinner.success
+  else
+    spinner.error(" Could not start service #{component_name.bold}")
+    puts "Debug information:\n\tComponent_name = #{component_name}\n\tDocker Output = #{output}"
+  end
+end
+
+def role_component(repos, component_name, skip_build = false)
```
Comment (j-maynard) [inline lib/components.rb:170]: role to roll

**Review by instantiator — APPROVED** [review]

**Pull request merged**